File: src/org/cappuccino_project/ide/intellij/plugin/stubs/impl/ObjJAccessorPropertyStubImpl.java
package org.cappuccino_project.ide.intellij.plugin.stubs.impl;

import com.intellij.psi.stubs.StubElement;
import org.cappuccino_project.ide.intellij.plugin.psi.impl.ObjJAccessorPropertyImpl;
import org.cappuccino_project.ide.intellij.plugin.psi.types.ObjJClassType;
import org.cappuccino_project.ide.intellij.plugin.stubs.interfaces.ObjJAccessorPropertyStub;
import org.cappuccino_project.ide.intellij.plugin.stubs.types.ObjJStubTypes;
import org.cappuccino_project.ide.intellij.plugin.psi.utils.ObjJMethodPsiUtils;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.Arrays;
import java.util.Collections;
import java.util.List;

public class ObjJAccessorPropertyStubImpl extends ObjJStubBaseImpl<ObjJAccessorPropertyImpl> implements ObjJAccessorPropertyStub {

    private final String containingClass;
    private final String variableName;
    private final String getter;
    private final String setter;
    private String varType;
    private final boolean shouldResolve;

    public ObjJAccessorPropertyStubImpl(final StubElement parent, final String containingClass,
                                        @Nullable String varType, @Nullable String variableName,
                                        @Nullable String getter, @Nullable String setter,
                                        final boolean shouldResolve
    ) {
        super(parent, ObjJStubTypes.ACCESSOR_PROPERTY);
        this.containingClass = containingClass;
        this.varType = varType != null && !varType.isEmpty() ? varType : null;
        this.variableName = variableName != null && !variableName.isEmpty() ? variableName : null;
        this.getter = getter != null && !getter.isEmpty() ? getter : null;
        this.setter = setter != null && !setter.isEmpty() ? setter : null;
        this.shouldResolve = shouldResolve;
    }

    @NotNull
    @Override
    public String getContainingClass() {
        return containingClass;
    }

    @Nullable
    @Override
    public String getVariableName() {
        return variableName;
    }

    @Nullable
    @Override
    public String getGetter() {
        return getter;
    }

    @javax.annotation.Nullable
    @Override
    public String getSetter() {
        return setter;
    }

    @Nullable
    @Override
    public String getVarType() {
        return varType;
    }

    @NotNull
    @Override
    public List<String> getParamTypes() {
        return getter != null ? Collections.emptyList() : Collections.singletonList(getVarType());
    }

    @NotNull
    @Override
    public List<String> getSelectorStrings() {
        return Arrays.asList(getSelectorString().split(ObjJMethodPsiUtils.SELECTOR_SYMBOL));
    }

    @NotNull
    @Override
    public String getSelectorString() {
        return getter != null ? getter : setter != null ? setter : ObjJMethodPsiUtils.EMPTY_SELECTOR;
    }

    @NotNull
    @Override
    public String getContainingClassName() {
        return containingClass;
    }

    @Override
    public boolean isRequired() {
        return false;
    }

    @NotNull
    @Override
    public ObjJClassType getReturnType() {
        return getter != null ? ObjJClassType.getClassType(varType) : ObjJClassType.VOID;
    }

    @NotNull
    @Override
    public String getReturnTypeAsString() {
        return getter != null ? varType : ObjJClassType.VOID_CLASS_NAME;
    }

    @Override
    public boolean isStatic() {
        return false;
    }

    @Override
    public boolean shouldResolve() {
        return shouldResolve;
    }
}

File: src/org/cappuccino_project/ide/intellij/plugin/stubs/impl/ObjJAccessorPropertyStubImpl.kt
package org.cappuccino_project.ide.intellij.plugin.stubs.impl

import com.intellij.psi.stubs.StubElement
import org.cappuccino_project.ide.intellij.plugin.psi.impl.ObjJAccessorPropertyImpl
import org.cappuccino_project.ide.intellij.plugin.psi.types.ObjJClassType
import org.cappuccino_project.ide.intellij.plugin.stubs.interfaces.ObjJAccessorPropertyStub
import org.cappuccino_project.ide.intellij.plugin.stubs.types.ObjJStubTypes
import org.cappuccino_project.ide.intellij.plugin.psi.utils.ObjJMethodPsiUtils

import java.util.Arrays
import java.util.Collections

class ObjJAccessorPropertyStubImpl(parent: StubElement<*>, override val containingClass: String,
                                   varType: String?, variableName: String?,
                                   getter: String?, setter: String?,
                                   private val shouldResolve: Boolean
) : ObjJStubBaseImpl<ObjJAccessorPropertyImpl>(parent, ObjJStubTypes.ACCESSOR_PROPERTY), ObjJAccessorPropertyStub {
    override val variableName: String?
    override val getter: String?
    override val setter: String?
    override val varType: String?

    override val paramTypes: List<String>
        get() = if (getter != null) emptyList() else listOf<String>(varType)

    override val selectorStrings: List<String>
        get() = Arrays.asList(*selectorString.split(ObjJMethodPsiUtils.SELECTOR_SYMBOL.toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray())

    override val selectorString: String
        get() = getter ?: (setter ?: ObjJMethodPsiUtils.EMPTY_SELECTOR)

    override val isRequired: Boolean
        get() = false

    override val returnType: ObjJClassType
        get() = if (getter != null) ObjJClassType.getClassType(varType) else ObjJClassType.VOID

    override val returnTypeAsString: String
        get() = if (getter != null) varType else ObjJClassType.VOID_CLASS_NAME

    override val isStatic: Boolean
        get() = false

    init {
        this.varType = if (varType != null && !varType.isEmpty()) varType else null
        this.variableName = if (variableName != null && !variableName.isEmpty()) variableName else null
        this.getter = if (getter != null && !getter.isEmpty()) getter else null
        this.setter = if (setter != null && !setter.isEmpty()) setter else null
    }

    override fun getContainingClassName(): String {
        return containingClass
    }

    override fun shouldResolve(): Boolean {
        return shouldResolve
    }
}
